canopen/pdo/base.py

```diff
@@ -9,13 +10,28 @@
 from canopen.sdo import SdoAbortedError
 from canopen import objectdictionary
 from canopen import variable
+try:
+    import canmatrix
+except ImportError:
+    canmatrix = None
 
 if TYPE_CHECKING:
     from canopen.network import Network
     from canopen import LocalNode, RemoteNode
     from canopen.pdo import RPDO, TPDO
     from canopen.sdo import SdoRecord
 
+
+def _disable_if(condition):
+    """Conditionally disable given function/method."""
+    def deco(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwds):
+            if not condition:
+                return func(*args, **kwds)
+        return wrapper
+    return deco
```

Comment:
Adding a decorator for this may seem overly verbose, but conditionally adding the method in other ways may mess up the documentation. I believe this is the best approach, but I'm open for other suggestions.

Reply:
Do we really need to disable (remove) the method?  My simple approach would be to just throw an exception when the import fails.  Currently that's an `ImportError`, but it could be wrapped with a better exception message explaining the optional dependency.

Reply:
Well, it's a convention to leave the method undefined if it's not supported, which is why I went for that, but other solutions are fine as long as they are documented.

Reply:
I'd favor a less intrusive solution. In fact, just letting the ImportError propagate, but documenting that it is possibly raised, would be sufficient.

Reply:
Sure, I'm fine with that.

Reply:
Addressed in 0b2eae97e1e259a802cb18498c525b85f268c486. Marking as resolved. PTAL.

Reply:
It _may_ be slightly harder to apply #358 to this API with the simpler approach. If the API is conditionally approached, the return annotation is pretty straight forward. Now, I'm not quite sure how to handle it (and I'm far from a typing expert). Perhaps a `-> "canmatrix.CanMatrix"` will work; I'm not sure about the best approach.

Reply:
cc. @sveinse, who probably knows more about typing than me.

Reply:
Using `-> "canmatrix.CanMatrix"` works fine (when `from __future__ import annotations` is used: `-> canmatrix.CanMatrix`) works find even if canmatrix is not available. It does not interfer with the run-time operations if it is not present.

However, the type checker do not like this construct where a module becomes a variable. The better fix is simply `pass` here.
```py
try:
    import canmatrix
except ImportError:
    canmatrix = None
```

In order to avoid errors by the type checker, canmatrix must be installed for it to check its types. I think it's fine to get the reported errors when it is not present.

Reply:
Thanks for chiming in. IIUC, the approach used as the PR stands now should not be too much of a hassle when typing is eventually added.

Reply:
My 2 cents on missing imports: Either
1. use conditional import in the top of the file and simply let it fail when calling the methods that require canmatrix. But help the user to understand that the raise exception is due to missing canmatrix
2. stick with importing inside the method. I know it's not pep8-like/pythonic, but the resulting `ImportError()` is very unambigous.

Reply:
Yeah. As I already wrote, I prefer something similar to my first proposal. Regarding the exception type, I don't have a strong opinion. I defer to the library maintainers :)

Reply:
> However, the type checker do not like this construct where a module becomes a variable. The better fix is simply pass here.

... or simply to set a (boolean) guard to be used, for example, in a decorator variant.

Reply:
Use of any other variable as sentinel is fine. Just not the same name as the module (`canmatrix`).

Reply:
Please do #2.  If you think a `RuntimeError` is not appropriate, we can manually raise an `ImportError`, which is very concise.  Who says an `ImportError` can only be raised at the time of importing? ;-)

But I'm not supposed to adding a new exception type either, if you think those ideas are too hacky.

Reply:
> Who says an ImportError can only be raised at the time of importing? ;-)

https://peps.python.org/pep-0690/ (too bad it didn't happen!) 😁 

None of the proposed ideas are too hacky; all is fine. It's just a matter of deciding on an approach and documenting it for the users convenience :)

Reply:
Regarding the exception type; the `TypeError` docs says this:

> This exception may be raised by user code to indicate that an attempted operation on an object is not supported, and is not meant to be. If an object is meant to support a given operation but has not yet provided an implementation, [NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError) is the proper exception to raise.

I guess we could use both `TypeError` and `NotImplementedError`, even though it might be a little bit unconventional. Perhaps `NotImplementedError` is the better choice.

Reply:
Then let's go for `NotImplementedError`.  Good idea!

Reply:
Let's continue the discussion in your more recent remarks. Marking this conversation as resolved.